### PR TITLE
Expose Transaction.Context.Request through the public agent API

### PR DESCRIPTION
**Pull request by gregkalapos**
This is a potential solution for https://github.com/elastic/apm-agent-dotnet/issues/124 

Some general thinking:
- We force users to provide required fields - aka: added constructors. It's very hard to have data that'd make the APM server unhappy.
- I added a POCO class `Api.Request`. The idea behind this is that I'd like to avoid exposing the model classes. Those can be changes (yes, breaking changes happen only on major versions), still I think it's not a good idea to expose those - most agents don't do it either.
- The `Request` class flattens the object hierarchy, it does not have properties that are other classes (like `Socket` and `Url`). Idea behind this: I think it's simply confusing... With that we would have an API where we constantly force users to new up objects and nullcheck things.
- Some agents call this `Context`. Since `Context` will probably disappear I'd simply not use that word - I'd not expose the concept of `Context` in generqal. To me this is just an additional information on the `ITransaction` called `Request`. 

- I'd use a similar solution with other things that we'd like to expose on `Context`.

### CODE REVIEW

File: src/Elastic.Apm/Api/Request.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+namespace Elastic.Apm.Api
+{
+	/// <summary>
+	/// Encapsulates Request related information that can be attached to an <see cref="ITransaction" />.
+	/// See <see cref="ITransaction.Request" />
+	/// </summary>
+	public class Request
```
Comment (gregkalapos) [inline src/Elastic.Apm/Api/Request.cs]: We have another internal class called `Request`, but that does not affect users and I think we can live with this.

File: src/Elastic.Apm/Model/Payload/Transaction.cs [hunk 2]
```diff
@@ -51,6 +53,21 @@ public Transaction(IApmLogger logger, string name, string type, IPayloadSender s
 
 		public string Name { get; set; }
 
+		public Api.Request Request
+		{
+			get => _request;
+			set
+			{
+				_request = value;
+				_context.Value.Request = new Request(value.Method,
```
Comment (gregkalapos) [inline src/Elastic.Apm/Model/Payload/Transaction.cs]: This can be also done during serialization... I'm not sure where to do this... I'm ok with moving this.

File: test/Elastic.Apm.AspNetCore.Tests/AspNetCoreMiddlewareTests.cs [hunk 3]
```diff
@@ -115,7 +115,7 @@ public async Task FailingRequestWithoutConfiguredExceptionPage()
 			Assert.Single(_capturedPayload.Errors[0].Errors);
 
 			//also make sure the tag is captured
-			Assert.Equal(((_capturedPayload.Errors[0] as Error)?.Errors[0] as Error.ErrorDetail)?.Context.Tags["foo"], "bar");
```
Comment (gregkalapos) [inline test/Elastic.Apm.AspNetCore.Tests/AspNetCoreMiddlewareTests.cs]: Unrelated change, sry about this.

File: test/Elastic.Apm.Tests/ApiTests/ConvenientApiTransactionTests.cs [hunk 4]
```diff
@@ -380,7 +381,7 @@ public async Task CancelledAsyncTask()
 			Assert.Equal(TransactionType, payloadSender.Payloads[0].Transactions[0].Type);
 
 			//`await Task.Delay` is inaccurate, so we enable 10% error in the test
-			var expectedTransactionLength = SleepLength - (SleepLength * 0.1);
```
Comment (gregkalapos) [inline test/Elastic.Apm.Tests/ApiTests/ConvenientApiTransactionTests.cs]: Unrelated change, sry about this.

File: src/Elastic.Apm.AspNetCore/ApmMiddleware.cs [hunk 5]
```diff
@@ -32,21 +34,20 @@ public async Task InvokeAsync(HttpContext context)
...
 
-			transaction.Context.Request = new Request
+			var url = new Url
+			{
+				Full = context.Request?.Path.Value,
+				HostName = context.Request.Host.Host,
+				Protocol = GetProtocolName(context.Request.Protocol),
+				Raw = context.Request?.Path.Value //TODO
+			};
+			transaction.Context.Request = new Payload.Request(context.Request.Method, url)
```
Comment (gregkalapos) [inline src/Elastic.Apm.AspNetCore/ApmMiddleware.cs]: In internal code we can still directly set things on `Context`. This may be a little bit confusing, but from perf reasons I think we should do it this way - we avoid copying stuff this way.

File: src/Elastic.Apm/Model/Payload/Transaction.cs (hunk 2 above)
Comment (gregkalapos) [reply src/Elastic.Apm/Model/Payload/Transaction.cs]: 🤦‍♂️  We have to... this won't work if we do it here.

Comment (gregkalapos) [reply src/Elastic.Apm/Model/Payload/Transaction.cs]: Done, moved this to `End()`

**Comment by skynet2** [comment]
Hi @gregkalapos ,

Are you going to expose Response ?

File: src/Elastic.Apm/Model/Payload/Transaction.cs [hunk 6]
```diff
@@ -78,6 +80,16 @@ public void End()
 		{
 			if (!Duration.HasValue) Duration = (DateTimeOffset.UtcNow - Start).TotalMilliseconds;
 
+			if (_context?.Value != null && _context.Value.Request == null && Request != null)
+			{
+				_context.Value.Request = new Request(Request.Method,
```
Comment (skynet2) [inline src/Elastic.Apm/Model/Payload/Transaction.cs]: Good idea

**Comment by gregkalapos** [comment]
> Hi @gregkalapos ,
> 
> Are you going to expose Response ?

Yes, definitely! The concept would be the same: we create a POCO object for the API surface and expose that on `ITransaction`.  Assuming this PR is accepted... 

So this PR is about 2 thing: 1) the general concept about how we expose these things and 2) the specific implementation for `Request`. Once this is done we can move on and implement this for more things.

File: src/Elastic.Apm/Api/Request.cs (hunk 1 above)
Comment (skynet2) [inline src/Elastic.Apm/Api/Request.cs]: Hi @gregkalapos , maybe you can add Body field to the object please ?

Comment (gregkalapos) [reply src/Elastic.Apm/Api/Request.cs]: Good catch, thanks! 

Fixed.

Comment (skynet2) [reply src/Elastic.Apm/Api/Request.cs]: Thanks @gregkalapos

**Comment by skynet2** [comment]
Hi @gregkalapos ,

Is there any known time when will we have this code in master branch?

**Comment by gregkalapos** [comment]
This got outvoted.

There will be a follow-up PR according to https://github.com/elastic/apm-agent-dotnet/issues/124, which will expose the internal server API.
